OpenRA.Mods.Common/Activities/Air/Fly.cs

```diff
@@ -166,6 +169,12 @@ public override bool Tick(Actor self)
 				return false;
 			}
 
+			// Stop if we have encountered an obstacle and are close enough.
+			if (aircraft.OldMovementTypes.HasMovementType(MovementType.Horizontal) &&
```

Comment:
> won't both OldMovementTypes and CurrentMovementTypes remain as true?

No. `CurrentMovementTypes` is already based on `CenterPosition`.

Reply:
Testing this ingame showed that this does have the problem I was worried about - build a group of e.g. 10 chinooks, and notice that they still behave quite badly at waypoints.

When I said compare `CenterPosition` I meant something like

```diff
diff --git a/OpenRA.Mods.Common/Activities/Air/Fly.cs b/OpenRA.Mods.Common/Activities/Air/Fly.cs
index a3f0a7be4a..99786598d1 100644
--- a/OpenRA.Mods.Common/Activities/Air/Fly.cs
+++ b/OpenRA.Mods.Common/Activities/Air/Fly.cs
@@ -28,6 +28,7 @@ public class Fly : Activity
                Target target;
                Target lastVisibleTarget;
                bool useLastVisibleTarget;
+               int? expectedDistanceToTarget = null;
 
                public Fly(Actor self, Target t, WDist nearEnough, WPos? initialTargetPosition = null, Color? targetLineColor = null)
                        : this(self, t, initialTargetPosition, targetLineColor)
@@ -175,6 +176,15 @@ public override bool Tick(Actor self)
                        }
 
                        // Stop if we have encountered an obstacle and are close enough.
+                       if (expectedDistanceToTarget.HasValue)
+                       {
+                               var distanceToTarget = (self.CenterPosition - checkTarget.CenterPosition).Length;
+
+                               // Something is stopping us from moving closer - abort if we are already near enough
+                               var blockedThreshold = aircraft.MovementSpeed * 2 / 3;
+                               if (distanceToTarget > expectedDistanceToTarget + blockedThreshold && delta.HorizontalLengthSquared <= nearEnough.LengthSquared)
+                                       return true;
+                       }
                        if (aircraft.OldMovementTypes.HasMovementType(MovementType.Horizontal) &&
                                !aircraft.CurrentMovementTypes.HasMovementType(MovementType.Horizontal) &&
                                delta.HorizontalLengthSquared <= nearEnough.LengthSquared)
@@ -229,6 +239,8 @@ public override bool Tick(Actor self)
 
                        FlyTick(self, aircraft, desiredFacing, aircraft.Info.CruiseAltitude);
 
+                       expectedDistanceToTarget = (self.CenterPosition - checkTarget.CenterPosition).Length;
+
                        return false;
                }
```

Which seems to work reasonably well.

Reply:
Just did some tests. 

This makes them stop way too soon, which makes it very difficult to herd a group of helicopters to where you want them to. The problem is that this makes them give up simply when they brush up against another helicopter even if they aren't really blocked. 

When I add a requirement to only give up when blocked for x consecutive ticks, they will never give up at all, so we are back where we started. 

In contrast, with the movementTypes based logic, they will always break off at some point, but often after some amount of jostling.

The thing is: It will never be quite acceptable until you add distributed end points into it al la #16883, so this is never gonna be more than a stopgap. As a way to catch edge-cases in #16883 the more conservative approach based on movementTypes makes more sense. I feel that the other approach will just get in the way there.

Reply:
Updated as discussed on IRC